model/incoming_webhook.go

```diff
@@ -63,6 +93,185 @@ func IncomingWebhookListFromJson(data io.Reader) []*IncomingWebhook {
+	o.HmacAlgorithm = ""
+	o.TimestampModel = StringInterface{}
+	o.HmacModel = StringInterface{}
+	o.SignedContentModel = StringArray{}
+}
+
+func ValidModelSplitter(s *StringInterface) bool {
+	// empty splitBy is valid
+	splitValue, isSet := s.ValidHeaderModel("SplitBy")
+	if !isSet {
+		delete(*s, "Index")
+		return true
+	}
+
+	idxValue, _ := s.ValidHeaderModel("Index")
+	if len(splitValue) > 0 {
+		if !(len(idxValue) > 0) {
+			return false
+		}
+		intIdx, err := strconv.ParseInt(idxValue, 10, 0)
+		if err != nil {
+			return false
+		} else if int(intIdx) > 5 || int(intIdx) < 0 {
+			return false
+		}
+	} else {
+		if len(idxValue) > 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s StringInterface) ValidHeaderModel(mapKey string) (string, bool) {
+	_, isSet := s[mapKey]
+	if headerPartStr, ok := s[mapKey].(string); ok {
+		return headerPartStr, isSet
+	}
+	return "", isSet
+}
+
+func (o *IncomingWebhook) IsValidModels() bool {
+	if !(o.HmacAlgorithm == "HMAC-SHA1" || o.HmacAlgorithm == "HMAC-SHA256") {
+		return false
+	}
+
+	if !(len(o.SignedContentModel) > 0 && len(o.SignedContentModel) < 6) {
+		return false
+	}
+
+	hmacModelHeader, ok := o.HmacModel["HeaderName"].(string)
+	if !(ok && 0 < len(hmacModelHeader) && len(hmacModelHeader) < 26) {
+		return false
+	}
+	if !ValidModelSplitter(&o.HmacModel) {
+		return false
+	}
+
+	if pre, pSet := o.HmacModel.ValidHeaderModel("Prefix"); pSet && pre == "" {
+		return false
+	}
+
+	timestampModelH := o.TimestampModel["HeaderName"]
+	if timestampModelH == nil {
+		o.TimestampModel = StringInterface{}
+	} else {
+		tsModelHeader, ok1 := timestampModelH.(string)
+		if !(ok1 && 0 < len(tsModelHeader) && len(tsModelHeader) < 26) {
+			return false
+
+		}
+
+		if !ValidModelSplitter(&o.TimestampModel) {
+			return false
+		}
+
+		if pre, pSet := o.TimestampModel.ValidHeaderModel("Prefix"); pSet && (pre == "" || len(pre) > 8) {
+			return false
+		}
+
+	}
+
+	return true
+}
+
+func ParseHeaderModel(s StringInterface) HeaderModel {
+	headerModel := HeaderModel{}
+	for _, modelKey := range [4]string{"HeaderName", "SplitBy", "Index", "Prefix"} {
+		if headerPart, ok := s[modelKey].(string); ok {
+			switch modelKey {
+			case "HeaderName":
+				headerModel.HeaderName = headerPart
+			case "SplitBy":
+				headerModel.SplitBy = headerPart
+			case "Index":
+				headerModel.Index = headerPart
+			case "Prefix":
+				headerModel.Prefix = headerPart
+			}
+		}
+	}
+	return headerModel
+}
+
+func (sp SignedIncomingHook) VerifySignature(secretToken string) bool {
+	var mac hash.Hash
+	if sp.Algorithm == "HMAC-SHA1" {
+		mac = hmac.New(sha1.New, []byte(secretToken))
+	} else {
+		mac = hmac.New(sha256.New, []byte(secretToken))
+	}
+
+	mac.Write(*sp.ContentToSign)
+	sha := mac.Sum(nil)
+
+	sig, err := hex.DecodeString(sp.Signature)
+	if err != nil {
+		return false
+	}
+
+	return hmac.Equal(sig, sha)
+}
+
+func (o *IncomingWebhook) ParseHeader(signedHook *SignedIncomingHook, r *http.Request) *AppError {
```

Comment:
A lot of this /feels/ like belongs in `app`, not `model`.

Reply:
These too, moved to web where the api receives webhook requests from senders.